### PR TITLE
tickets/PIPE2D-1136: Use an uncontaminated part of image to estimate the dark current

### PR DESCRIPTION
**Pull request by RobertLuptonTheGood**
Some CCDs (e.g. r3) have a time-variable glow, so you can't estimate the dark current from the darktime.  Instead, we estimate it from a signal-free portion of the image, as specified using isr.darkBBoxes; we clip to avoid the nefarious effects of cosmic rays.

### CODE REVIEW

File: python/lsst/obs/pfs/isrTask.py [hunk 1]
```diff
@@ -131,6 +132,17 @@ class PfsIsrTaskConfig(ipIsr.IsrTaskConfig):
     windowed = pexConfig.Field(dtype=bool, default=False,
                                doc="Reduction of windowed data, for real-time acquisition? Implies "
                                "overscanFitType=MEDIAN")
+    darkBBoxes = pexConfig.DictField(
+        keytype=str, itemtype=str,
+        default=dict(r3="[geom.BoxI(geom.PointI(4045, 2660), geom.PointI(4095, 2910))]"),
+        doc="List of BBoxes specifying pixels used to determine amplitude of dark frame")
+    fitDarkClipPercentiles = pexConfig.ListField(dtype=float, default=[0, 5],
+                                                 doc="""\
```
Comment (PaulPrice) [inline python/lsst/obs/pfs/isrTask.py:140]: Shouldn't need the backslash here, right?

File: python/lsst/obs/pfs/isrTask.py [hunk 2]
```diff
@@ -151,6 +163,15 @@ def validate(self):
         if self.doApplyGains:
             raise RuntimeError("doApplyGains must be False: gains are applied automatically.")
 
+        for detName, bboxStr in self.darkBBoxes.items():
+            try:
+                eval(bboxStr)           # can't set self.darkBBoxes[detName] here as it fails type validation
+                estr = ""
+            except Exception as e:
+                estr = str(e)           # it's clearer not to raise within the try block
```
Comment (PaulPrice) [inline python/lsst/obs/pfs/isrTask.py:171]: Not disagreeing, but why do you think it's clearer?

File: python/lsst/obs/pfs/isrTask.py [hunk 3]
```diff
@@ -622,6 +660,61 @@ def maskAmplifier(self, exposure, amp, defects):
...
+           Specify which pixels should be used to estimate amplitude
+        data : `lsst.afw.image.MaskedImage`
+           Image to process.  The image is modified by this method.
+        dark : `lsst.afw.image.MaskedImage`
+            Dark image of the same size as ``data``.
+        clipPercentiles : `list` of `float`
+           Percentiles at which we clip the data in each iteration
+
+        Raises
+        ------
```
Comment (PaulPrice) [inline python/lsst/obs/pfs/isrTask.py:681]: Need to finish docstring.

File: python/lsst/obs/pfs/isrTask.py [hunk 4]
```diff
@@ -622,6 +660,61 @@ def maskAmplifier(self, exposure, amp, defects):
 
         return badAmp
 
+    def darkCorrectionFromBBoxes(self, bboxes, data, dark, clipPercentiles):
```
Comment (PaulPrice) [inline python/lsst/obs/pfs/isrTask.py:663]: Since this is a method, should drop `clipPercentiles`, and simply take them from `self.config`.

**Review by PaulPrice — COMMENTED** [review]
Code looks good, but I think we need to discuss versioning the `darkBBoxes`.

File: python/lsst/obs/pfs/isrTask.py (hunk 1 above)
Comment (RobertLuptonTheGood) [reply python/lsst/obs/pfs/isrTask.py:140]: Otherwise there's a leading `\n`

File: python/lsst/obs/pfs/isrTask.py (hunk 2 above)
Comment (RobertLuptonTheGood) [reply python/lsst/obs/pfs/isrTask.py:171]: Because it avoids the double traceback ("exception while processing exception" or something like that)

Comment (PaulPrice) [reply python/lsst/obs/pfs/isrTask.py:171]: I've never thought of that. Does `raise ValueError("...") from e` solve the problem?

Comment (RobertLuptonTheGood) [reply python/lsst/obs/pfs/isrTask.py:171]: e doesn't exist outside the except block, and raising it in the block doesn't help.   But I clarified the comment

**Pull request merged**